File: client/src/lib/api/meet.ts
/**
 * Meet API Module
 *
 * Migrated to use API Factory pattern.
 * @see factory.ts for client creation details
 */
import { getClient, getServiceBaseUrl, ServiceName } from './factory';

// Get the meet service client (cached)
const meetClient = getClient(ServiceName.MEET);
const MEET_BASE_URL = getServiceBaseUrl(ServiceName.MEET);

// ============================================================================
// Types
// ============================================================================

export interface Room {
    id: string;
    name: string;
    description?: string;
    room_code: string;
    status: 'scheduled' | 'active' | 'ended';
    is_private: boolean;
    max_participants?: number;
    scheduled_start?: string;
    scheduled_end?: string;
    actual_start?: string;
    actual_end?: string;
    settings?: Record<string, unknown>;
    created_at: string;
    participant_count?: number;
    livekit_url?: string;
}

export interface CreateRoomRequest {
    name: string;
    description?: string;
    is_private?: boolean;
    password?: string;
    max_participants?: number;
    scheduled_start?: string;
    scheduled_end?: string;
    settings?: Record<string, unknown>;
}

export interface UpdateRoomRequest {
    name?: string;
    description?: string;
    is_private?: boolean;
    password?: string;
    max_participants?: number;
    scheduled_start?: string;
    scheduled_end?: string;
    settings?: Record<string, unknown>;
}

export interface Participant {
    id: string;
    user_id?: string;
    display_name: string;
    role: 'host' | 'moderator' | 'participant';
    joined_at: string;
    is_muted: boolean;
    is_video_off: boolean;
    is_screen_sharing: boolean;
}

export interface TokenResponse {
    token: string;
    livekit_url: string;
    room_name: string;
}

export interface Recording {
    id: string;
    room_id: string;
    status: 'recording' | 'processing' | 'ready' | 'failed';
    started_at: string;
    ended_at?: string;
    duration_seconds?: number;
    file_size_bytes?: number;
    download_url?: string;
}

/** Response of `GET /meet/rooms/by-code/:code/recording` (public). */
export interface ActiveRecordingByCode {
    is_recording: boolean;
    recording_id?: string;
    started_at?: string;
    room_id: string;
}

export interface MeetingHistory {
    id: string;
    room_name: string;
    started_at: string;
    ended_at?: string;
    duration_seconds?: number;
    participant_count: number;
    had_recording: boolean;
}

export interface MeetConfig {
    livekit_url: string;
    max_participants_per_room: number;
    recording_enabled: boolean;
}

export interface MuteRequest {
    audio?: boolean;
    video?: boolean;
}

export interface InstantRoomResponse {
    id: string;
    room_code: string;
    name: string;
    livekit_url?: string;
}

export interface LobbyInfo {
    /** Room UUID (for code→id resolution without a separate round-trip). */
    room_id: string;
    /** Human-readable room name. */
    room_name: string;
    /** Whether the room accepts joins (not `ended`). */
    is_open: boolean;
    /** Whether the host requires explicit admission (knock flow). */
    requires_knock: boolean;
    /** Whether the room is password-protected. */
    has_password: boolean;
}

export interface KnockRequest {
    display_name: string;
    identity?: string;
}

export type KnockStatus = 'pending' | 'admitted' | 'denied';

export interface KnockResponse {
    request_id: string;
    identity: string;
    status: KnockStatus;
}

export interface KnockEntry {
    request_id: string;
    identity: string;
    display_name: string;
    status: KnockStatus;
    created_at: string;
    resolved_at?: string;
}

export interface KnockStatusResponse {
    status: KnockStatus;
}

export interface JoinRoomResponse {
    token: string;
    livekit_url: string;
    room_name: string;
    room_code: string;
}

// ============================================================================
// Meet API
// ============================================================================

export const meetApi = {
    // ========================================================================
    // Config
    // ========================================================================

    getConfig: () =>
        meetClient.get<MeetConfig>('/meet/config'),

    // ========================================================================
    // Rooms
    // ========================================================================

    listRooms: () =>
        meetClient.get<Room[]>('/meet/rooms'),

    getRoom: (id: string) =>
        meetClient.get<Room>(`/meet/rooms/${id}`),

    createRoom: (data: CreateRoomRequest) =>
        meetClient.post<Room>('/meet/rooms', data),

    createInstantRoom: () =>
        meetClient.post<InstantRoomResponse>('/meet/rooms/instant'),

    getLobby: (code: string) =>
        meetClient.get<LobbyInfo>(`/meet/rooms/${code}/lobby`),

    /** DB-backed knock flow (persisted in meet.waiting_room_requests). */
    knock: (code: string, data: KnockRequest) =>
        meetClient.post<KnockResponse>(`/meet/rooms/${code}/knock`, data),

    /** Public polling endpoint — returns the current status for an identity. */
    getKnockStatus: (code: string, identity: string) =>
        meetClient.get<KnockStatusResponse>(
            `/meet/rooms/${code}/knock-status?identity=${encodeURIComponent(identity)}`,
        ),

    /** Host-only — returns pending knockers for this room. */
    listKnocks: (code: string) =>
        meetClient.get<KnockEntry[]>(`/meet/rooms/${code}/knocks`),

    admitKnock: (code: string, identity: string) =>
        meetClient.post<KnockEntry>(
            `/meet/rooms/${code}/admit/${encodeURIComponent(identity)}`,
        ),

    denyKnock: (code: string, identity: string) =>
        meetClient.post<KnockEntry>(
            `/meet/rooms/${code}/deny/${encodeURIComponent(identity)}`,
        ),

    joinByCode: (code: string, displayName?: string) =>
        meetClient.post<JoinRoomResponse>(`/meet/rooms/${code}/join`, {
            display_name: displayName,
        }),

    updateRoom: (id: string, data: UpdateRoomRequest) =>
        meetClient.put<Room>(`/meet/rooms/${id}`, data),

    deleteRoom: (id: string) =>
        meetClient.delete(`/meet/rooms/${id}`),

    endRoom: (id: string) =>
        meetClient.post(`/meet/rooms/${id}/end`),

    // ========================================================================
    // Tokens
    // ========================================================================

    getToken: () =>
        meetClient.get<TokenResponse>('/meet/token'),

    getRoomToken: (roomId: string) =>
        meetClient.get<TokenResponse>(`/meet/rooms/${roomId}/token`),

    // ========================================================================
    // Participants
    // ========================================================================

    listParticipants: (roomId: string) =>
        meetClient.get<Participant[]>(`/meet/rooms/${roomId}/participants`),

    kickParticipant: (roomId: string, userId: string) =>
        meetClient.post(`/meet/rooms/${roomId}/participants/${userId}/kick`),

    muteParticipant: (roomId: string, userId: string, data: MuteRequest) =>
        meetClient.post(`/meet/rooms/${roomId}/participants/${userId}/mute`, data),

    // ========================================================================
    // Recordings
    // ========================================================================

    listRecordings: (roomId: string) =>
        meetClient.get<Recording[]>(`/meet/rooms/${roomId}/recordings`),

    startRecording: (roomId: string) =>
        meetClient.post<Recording>(`/meet/rooms/${roomId}/recordings`),

    getRecording: (recordingId: string) =>
        meetClient.get<Recording>(`/meet/recordings/${recordingId}`),

    stopRecording: (recordingId: string) =>
        meetClient.post<Recording>(`/meet/recordings/${recordingId}/stop`),

    deleteRecording: (recordingId: string) =>
        meetClient.delete(`/meet/recordings/${recordingId}`),

    // Code-addressed convenience endpoints (host-only except the first).
    recordings: {
        /** Public. Returns `{ is_recording, recording_id?, started_at?, room_id }`. */
        getActiveByCode: (code: string) =>
            meetClient.get<ActiveRecordingByCode>(
                `/meet/rooms/by-code/${encodeURIComponent(code)}/recording`,
            ),
        /** Host-only. Starts a DB-tracked recording for the given room code. */
        startByCode: (code: string) =>
            meetClient.post<Recording>(
                `/meet/rooms/by-code/${encodeURIComponent(code)}/recording/start`,
            ),
        /** Host-only. Stops the active recording (if any) for the given room code. */
        stopByCode: (code: string) =>
            meetClient.post<Recording>(
                `/meet/rooms/by-code/${encodeURIComponent(code)}/recording/stop`,
            ),
        listByRoom: (roomId: string) =>
            meetClient.get<Recording[]>(`/meet/rooms/${roomId}/recordings`),
    },

    // ========================================================================
    // History
    // ========================================================================

    listHistory: () =>
        meetClient.get<MeetingHistory[]>('/meet/history'),

    // ========================================================================
    // Voicemails — /api/v1/meet/voicemails
    // ========================================================================

    voicemails: {
        list: () =>
            meetClient.get<Voicemail[]>('/meet/voicemails'),
        delete: (id: string) =>
            meetClient.delete(`/meet/voicemails/${id}`),
        markRead: (id: string) =>
            meetClient.patch(`/meet/voicemails/${id}/read`),
    },

    // ========================================================================
    // Video Messages — /api/v1/meet/video-messages
    // ========================================================================

    videoMessages: {
        list: () =>
            meetClient.get<VideoMessage[]>('/meet/video-messages'),
        create: (data: CreateVideoMessageRequest) =>
            meetClient.post<VideoMessage>('/meet/video-messages', data),
        delete: (id: string) =>
            meetClient.delete(`/meet/video-messages/${id}`),
        markRead: (id: string) =>
            meetClient.patch(`/meet/video-messages/${id}/read`),
    },

    // ========================================================================
    // Live transcription (Phase 3b)
    // ========================================================================

    transcription: {
        /**
         * Persist a single utterance. Called by every client with the
         * transcription toggle enabled, once per audio chunk successfully
         * transcribed by signapps-media.
         */
        ingest: (code: string, data: TranscriptionChunk) =>
            meetClient.post<TranscriptionIngestResponse>(
                `/meet/rooms/${encodeURIComponent(code)}/transcription/ingest`,
                data,
            ),
        /** Fetch recent utterances (ordered by `timestamp_ms ASC`). */
        history: (code: string, limit = 200) =>
            meetClient.get<TranscriptionEntry[]>(
                `/meet/rooms/${encodeURIComponent(code)}/transcription/history?limit=${limit}`,
            ),
        /** Absolute URL for the export endpoint (for anchor hrefs / `window.open`). */
        exportUrl: (code: string, format: TranscriptionExportFormat = 'txt') =>
            `${MEET_BASE_URL}/meet/rooms/${encodeURIComponent(
                code,
            )}/transcription/export?format=${format}`,
    },

    // ========================================================================
    // Raise hand (Phase 3c)
    // ========================================================================

    hands: {
        list: (code: string) =>
            meetClient.get<RaisedHand[]>(
                `/meet/rooms/${encodeURIComponent(code)}/hands`,
            ),
        raise: (code: string) =>
            meetClient.post<RaiseHandResponse>(
                `/meet/rooms/${encodeURIComponent(code)}/hands/raise`,
            ),
        lower: (code: string) =>
            meetClient.post<LowerHandResponse>(
                `/meet/rooms/${encodeURIComponent(code)}/hands/lower`,
            ),
        lowerOther: (code: string, identity: string) =>
            meetClient.post<LowerHandResponse>(
                `/meet/rooms/${encodeURIComponent(code)}/hands/lower/${encodeURIComponent(
                    identity,
                )}`,
            ),
    },

    // ========================================================================
    // Polls (Phase 3c)
    // ========================================================================

    polls: {
        list: (code: string) =>
            meetClient.get<MeetPoll[]>(
                `/meet/rooms/${encodeURIComponent(code)}/polls`,
            ),
        create: (code: string, data: CreateMeetPollRequest) =>
            meetClient.post<MeetPoll>(
                `/meet/rooms/${encodeURIComponent(code)}/polls`,
                data,
            ),
        vote: (pollId: string, data: VoteMeetPollRequest) =>
            meetClient.post<MeetPoll>(
                `/meet/polls/${encodeURIComponent(pollId)}/vote`,
                data,
            ),
        close: (pollId: string) =>
            meetClient.post<MeetPoll>(
                `/meet/polls/${encodeURIComponent(pollId)}/close`,
            ),
    },
};

// ============================================================================
// Meet — Voicemail & VideoMessage types
// ============================================================================

export interface Voicemail {
    id: string;
    user_id: string;
    caller_name?: string;
    caller_phone?: string;
    duration_seconds?: number;
    transcription?: string;
    audio_storage_key?: string;
    is_read: boolean;
    created_at: string;
}

export interface VideoMessage {
    id: string;
    sender_id: string;
    recipient_id: string;
    duration_seconds?: number;
    thumbnail_url?: string;
    video_storage_key?: string;
    is_read: boolean;
    created_at: string;
}

export interface CreateVideoMessageRequest {
    recipient_id: string;
    duration_seconds?: number;
    thumbnail_url?: string;
    video_storage_key?: string;
}

// ============================================================================
// Live transcription (Phase 3b)
// ============================================================================

/** Payload for `POST /meet/rooms/:code/transcription/ingest` and the
 * LiveKit data-channel broadcast — identical shape on both sides. */
export interface TranscriptionChunk {
    speaker_identity: string;
    text: string;
    timestamp_ms: number;
    language?: string | null;
}

/** Response of `POST /meet/rooms/:code/transcription/ingest`. */
export interface TranscriptionIngestResponse {
    id: string;
}

/** One persisted utterance returned by the history endpoint. */
export interface TranscriptionEntry {
    id: string;
    room_id: string;
    speaker_identity: string;
    text: string;
    timestamp_ms: number;
    language?: string | null;
    created_at: string;
}

export type TranscriptionExportFormat = 'md' | 'srt' | 'txt';

// ============================================================================
// Raise hand (Phase 3c)
// ============================================================================

export interface RaisedHand {
    identity: string;
    raised_at: string;
}

export interface RaiseHandResponse {
    identity: string;
    raised_at: string;
}

export interface LowerHandResponse {
    identity: string;
}

// ============================================================================
// Polls (Phase 3c)
// ============================================================================

/** A live poll inside a meet room. Named `MeetPoll` to avoid a name
 * collision with the calendar module's `Poll` type. */
export interface MeetPoll {
    id: string;
    room_id: string;
    created_by: string;
    question: string;
    /** Options is a JSON array of strings. */
    options: string[];
    /** Vote map keyed by identity. */
    votes: Record<string, number>;
    created_at: string;
    closed_at?: string | null;
}

export interface CreateMeetPollRequest {
    question: string;
    options: string[];
}

export interface VoteMeetPollRequest {
    option_index: number;
}

File: services/signapps-meet/src/models/meet.rs
//! Meet service models

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

/// A meeting room
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Room {
    pub id: Uuid,
    pub name: String,
    pub description: Option<String>,
    pub created_by: Uuid,
    pub room_code: String,
    pub status: String, // "scheduled", "active", "ended"
    pub is_private: bool,
    pub password_hash: Option<String>,
    pub max_participants: Option<i32>,
    pub scheduled_start: Option<DateTime<Utc>>,
    pub scheduled_end: Option<DateTime<Utc>>,
    pub actual_start: Option<DateTime<Utc>>,
    pub actual_end: Option<DateTime<Utc>>,
    pub settings: Option<serde_json::Value>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// Room participant tracking
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct RoomParticipant {
    pub id: Uuid,
    pub room_id: Uuid,
    pub user_id: Option<Uuid>,
    pub display_name: String,
    pub role: String, // "host", "moderator", "participant"
    pub joined_at: DateTime<Utc>,
    pub left_at: Option<DateTime<Utc>>,
    pub is_muted: bool,
    pub is_video_off: bool,
    pub is_screen_sharing: bool,
}

/// Recording metadata
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Recording {
    pub id: Uuid,
    pub room_id: Uuid,
    pub started_by: Uuid,
    pub status: String, // "recording", "processing", "ready", "failed"
    pub started_at: DateTime<Utc>,
    pub ended_at: Option<DateTime<Utc>>,
    pub duration_seconds: Option<i32>,
    pub file_size_bytes: Option<i64>,
    pub storage_path: Option<String>,
    pub storage_bucket: Option<String>,
    pub created_at: DateTime<Utc>,
}

/// Meeting history entry (for analytics)
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct MeetingHistory {
    pub id: Uuid,
    pub room_id: Uuid,
    pub room_name: String,
    pub host_id: Uuid,
    pub started_at: DateTime<Utc>,
    pub ended_at: Option<DateTime<Utc>>,
    pub duration_seconds: Option<i32>,
    pub participant_count: i32,
    pub max_concurrent_participants: i32,
    pub had_recording: bool,
    pub had_screen_share: bool,
}

// ============================================================================
// Request/Response DTOs
// ============================================================================

#[derive(Debug, Deserialize, utoipa::ToSchema)]
/// Request payload for CreateRoom operation.
pub struct CreateRoomRequest {
    pub name: String,
    pub description: Option<String>,
    pub is_private: Option<bool>,
    pub password: Option<String>,
    pub max_participants: Option<i32>,
    pub scheduled_start: Option<DateTime<Utc>>,
    pub scheduled_end: Option<DateTime<Utc>>,
    pub settings: Option<serde_json::Value>,
}

#[derive(Debug, Deserialize, utoipa::ToSchema)]
/// Request payload for UpdateRoom operation.
pub struct UpdateRoomRequest {
    pub name: Option<String>,
    pub description: Option<String>,
    pub is_private: Option<bool>,
    pub password: Option<String>,
    pub max_participants: Option<i32>,
    pub scheduled_start: Option<DateTime<Utc>>,
    pub scheduled_end: Option<DateTime<Utc>>,
    pub settings: Option<serde_json::Value>,
}

#[derive(Debug, Serialize, utoipa::ToSchema)]
/// Response payload for Room operation.
pub struct RoomResponse {
    pub id: Uuid,
    pub name: String,
    pub description: Option<String>,
    pub room_code: String,
    pub status: String,
    pub is_private: bool,
    pub max_participants: Option<i32>,
    pub scheduled_start: Option<DateTime<Utc>>,
    pub scheduled_end: Option<DateTime<Utc>>,
    pub actual_start: Option<DateTime<Utc>>,
    pub actual_end: Option<DateTime<Utc>>,
    pub settings: Option<serde_json::Value>,
    pub created_at: DateTime<Utc>,
    pub participant_count: i32,
    pub livekit_url: String,
}

#[derive(Debug, Deserialize, utoipa::ToSchema)]
/// Request payload for JoinRoom operation.
pub struct JoinRoomRequest {
    pub password: Option<String>,
    pub display_name: Option<String>,
}

#[derive(Debug, Serialize, utoipa::ToSchema)]
/// Response payload for Token operation.
pub struct TokenResponse {
    pub token: String,
    pub livekit_url: String,
    pub room_name: String,
}

#[derive(Debug, Serialize, utoipa::ToSchema)]
/// Response payload for Participant operation.
pub struct ParticipantResponse {
    pub id: Uuid,
    pub user_id: Option<Uuid>,
    pub display_name: String,
    pub role: String,
    pub joined_at: DateTime<Utc>,
    pub is_muted: bool,
    pub is_video_off: bool,
    pub is_screen_sharing: bool,
}

#[derive(Debug, Deserialize, utoipa::ToSchema)]
/// Request payload for Mute operation.
pub struct MuteRequest {
    pub audio: Option<bool>,
    pub video: Option<bool>,
}

#[derive(Debug, Serialize, utoipa::ToSchema)]
/// Response payload for Recording operation.
pub struct RecordingResponse {
    pub id: Uuid,
    pub room_id: Uuid,
    pub status: String,
    pub started_at: DateTime<Utc>,
    pub ended_at: Option<DateTime<Utc>>,
    pub duration_seconds: Option<i32>,
    pub file_size_bytes: Option<i64>,
    pub download_url: Option<String>,
}

#[derive(Debug, Serialize, utoipa::ToSchema)]
/// Response payload for MeetingHistory operation.
pub struct MeetingHistoryResponse {
    pub id: Uuid,
    pub room_name: String,
    pub started_at: DateTime<Utc>,
    pub ended_at: Option<DateTime<Utc>>,
    pub duration_seconds: Option<i32>,
    pub participant_count: i32,
    pub had_recording: bool,
}

#[derive(Debug, Serialize, utoipa::ToSchema)]
/// Response payload for Config operation.
pub struct ConfigResponse {
    pub livekit_url: String,
    pub max_participants_per_room: i32,
    pub recording_enabled: bool,
}
